dex/elrond_dex_farm/src/lib.rs

```diff
@@ -343,22 +368,92 @@ pub trait Farm:
+        let farm_token_id = self.farm_token_id().get();
+        require!(payment_token_id == farm_token_id, "Unknown farm token");
+
+        let farming_token = self.farming_token_id().get();
+        let reward_token = self.reward_token_id().get();
+        require!(
+            farming_token == reward_token,
+            "Farming token differ from reward token"
+        );
+        self.generate_aggregated_rewards(&reward_token);
+
+        let current_rps = self.reward_per_share().get();
+        let farm_attributes = self.get_farm_attributes(&payment_token_id, payment_token_nonce)?;
+        let reward = self.calculate_reward(
+            &payment_amount,
+            &current_rps,
+            &farm_attributes.reward_per_share,
+        );
+
+        if reward > 0 {
+            self.decrease_reward_reserve(&reward)?;
+        }
+
+        let farm_token_id = self.farm_token_id().get();
+        let new_farm_contribution = &payment_amount
+            + &(&reward * &Self::BigUint::from(farm_attributes.apr_multiplier as u64));
+
+        let new_initial_farming_amount = &(&farm_attributes.initial_farming_amount
+            * &payment_amount)
+            / &farm_attributes.current_farm_amount;
+        let new_reward_amount = &(&farm_attributes.compounded_reward * &payment_amount)
+            / &farm_attributes.current_farm_amount
+            + reward;
+        require!(
+            new_initial_farming_amount != 0,
+            "Farming token amount is zero"
+        );
+
+        let new_attributes = FarmTokenAttributes {
+            reward_per_share: current_rps,
+            entering_epoch: farm_attributes.entering_epoch,
+            apr_multiplier: farm_attributes.apr_multiplier,
+            with_locked_rewards: farm_attributes.with_locked_rewards,
+            initial_farming_amount: new_initial_farming_amount,
+            compounded_reward: new_reward_amount,
+            current_farm_amount: new_farm_contribution.clone(),
+        };
+
+        self.burn_farm_tokens(&farm_token_id, payment_token_nonce, &payment_amount)?;
+        let new_nonce =
+            self.create_farm_tokens(&new_farm_contribution, &farm_token_id, &new_attributes);
+        self.send_nft_tokens(
+            &farm_token_id,
+            new_nonce,
+            &new_farm_contribution,
+            &self.blockchain().get_caller(),
+            &opt_accept_funds_func,
+        );
+
+        Ok(GenericEsdtAmountPair {
+            token_id: farm_token_id,
+            token_nonce: new_nonce,
+            amount: new_farm_contribution,
+        })
+    }
+
     fn send_back_farming_tokens(
         &self,
         farming_token_id: &TokenIdentifier,
-        farming_amount: &mut Self::BigUint,
-        apr_multiplier: u8,
+        farming_amount: &Self::BigUint,
         destination: &Address,
-        with_locked_rewards: bool,
         opt_accept_funds_func: &OptionalArg<BoxedBytes>,
     ) -> SCResult<()> {
-        if with_locked_rewards {
```

Comment:
why is this code missing in the new implementation?

Reply:
before, the farming token amount one used to enter a farm were either 1:1 or 1:2 with the farm token amount. now, there is no longer the case and we use a variable in the token attributes so we know exactly the amount one used to enter farm. so no need for other calculations here since we already have the value we want to send back